tests/failpoints/test_engine.rs

```diff
@@ -516,3 +517,69 @@ fn test_concurrent_write_perf_context() {
         assert_ne!(old.apply_duration, new.apply_duration);
     }
 }
+
+#[test]
+fn test_recycle_with_stale_logbatch_at_tail() {
+    let dir = tempfile::Builder::new()
+        .prefix("test_recycle_with_stale_log_batch_at_tail")
+        .tempdir()
+        .unwrap();
+    let data = vec![b'x'; 1024];
+    let rid = 1;
+    let cfg_err = Config {
+        dir: dir.path().to_str().unwrap().to_owned(),
+        target_file_size: ReadableSize::kb(2),
+        purge_threshold: ReadableSize::kb(4),
+        enable_log_recycle: true,
+        ..Default::default()
+    };
+    // Force enable_log_recycle with Version::V1.
+    let _f = FailGuard::new("pipe_log::version::force_enable", "return");
+    // Force setting signature with Version::V1.
+    let _f = FailGuard::new("pipe_log::version::force_get_signature", "return");
+    {
+        let v1 = LogFileContext::new(
+            FileId {
+                seq: 10,
+                queue: LogQueue::Append,
+            },
+            Version::V1,
+        );
+        let v2 = LogFileContext::new(
+            FileId {
+                seq: 10,
+                queue: LogQueue::Append,
+            },
+            Version::V2,
+        );
+        assert!(v1.get_signature().is_none());
+        assert_eq!(v2.get_signature().unwrap() as u64, v2.id.seq);
+    }
+    // Do not truncate the active_file when exit
+    let _f = FailGuard::new("file_pipe_log::log_file_writer::skip_truncate", "return");
+    // assert_eq!(cfg_err.recycle_capacity(), 1);
+    assert_eq!(cfg_err.format_version, Version::V1);
+    let engine = Engine::open(cfg_err.clone()).unwrap();
+    append(&engine, rid, 1, 2, Some(&data)); // file_seq: 1
+    append(&engine, rid, 2, 3, Some(&data));
+    append(&engine, rid, 3, 4, Some(&data)); // file_seq: 2
+    append(&engine, rid, 4, 5, Some(&data));
+    append(&engine, rid, 5, 6, Some(&data)); // file_seq: 3
+    let append_first = engine.file_span(LogQueue::Append).0;
+    engine.compact_to(rid, 3);
+    engine.purge_expired_files().unwrap();
+    assert!(engine.file_span(LogQueue::Append).0 > append_first);
+    // append, written into seq: 3
+    append(&engine, rid, 4, 5, Some(&data));
+    // recycle, written into seq: 1
+    append(&engine, rid, 5, 6, Some(&data));
+    drop(engine);
+    // Recover the engine with invalid Version::default().
+    // Causing the final log file is a recycled file, containing rewritten
+    // LogBatchs and end with stale LogBatchs, `Engine::open(...)` should
+    // `panic` when recovering the relate `Memtable`.
+    assert!(catch_unwind_silent(|| {
+        Engine::open(cfg_err).unwrap();
```

Comment:
If you test `Engine::open(cfg_err);`, you are not testing the panic inside memtable, but the panic when unwrapping an error.
```suggestion
        Engine::open(cfg_err);
```

Reply:
Nop, the reason has been described as the above comment showed. I need to check the definition of `keep_data` again.

Reply:
These two comments are irrelevant. Your old code still passes if `Engine::open` returns an error (and unwrap panics), but it is supposed to catch a panic inside `Engine::open`.